### DISCUSSION TITLE
[Unknown Mod][Heroes] Hero Movement

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#10248] [General]
**Edit 2022/03/20.**

This can be summarized in two options:
- "heroes: hero has no diagonal movement penalty when he performs an action without leaving the square where he is located"
When the hero performs an action without moving from the square where he is located (collecting resources, collecting an artifact, attacking a hero, attacking wandering creatures, etc.) the diagonal movement penalty is not applied (so it is the same movement cost as if the hero performed the action horizontally or vertically, the terrain penalty where he is located being taken into account).
- "heroes: hero's predetermined path passes through the interactive objects to reduce the distance traveled"
The prediction of the hero's path always goes through the interactive objects of the adventure map to display and follow the shortest possible path.

-----

Version 0.8 fheroes2_windows_x64_SDL2
Map: 0Test10.MP2

Move the hero to the observation tower.
Select the gold stack as destination with the mouse cursor.
The program displays a default path for the hero.

![](https://i.imgur.com/84OKvds.png)

Click on the gold stack with the mouse cursor to accept the default path.
Do the same for all other resource stacks.
At the end of his movement, the hero will not be able to pick up the pile of mercury.

![](https://i.imgur.com/fRlvSkt.png)

Reload the scenario.
Move the hero to the observation tower.
Move the hero so that he is positioned right next to the stack of gold to pick it up in a horizontal motion.

![](https://i.imgur.com/YilJ3CR.png)

Do the same for all other resource stacks.
At the end of his movement, the hero will have collected all the resource piles and he will have been able to move one step south just next to the trees.

![](https://i.imgur.com/hfD0peZ.png)

This behavior with the default path with the mouse causes heroes to lose a considerable number of movement points during a game.
This should be enhanced.

One way to improve this would be that when a hero takes a resource, etc. diagonally (when the hero does not move to the square where the resource is located), he should only spend 1 movement (100 points for grass terrain for example) instead of 1.5 movement (150 points), i.e. always 1 movement (100 points).
This would avoid having to position the hero next to the resource to take it vertically or horizontally and thus conserve the 0.5 movement points, very annoying having to do this to optimize the movement.
And this also when the hero is positioned just in an adjacent square of a wandering creature he wants to attack.

Or even better, going further, movement points are returned to the hero if he could have collected multiple resources more efficiently during his movement for the current turn, so the player can concentrate on collecting resources instead than to think about optimizing the movement of its heroes.

The program automatically tests all the possibilities to collect the resources that the hero collected during his move and restores movement points to the hero if it was possible to do better (for the heroes of the AI too).
It is not necessary to change the visual of the hero's path, the movement points are restored in the background and the player does not even notice it.

Another case with the hero movement.

Reload the scenario.
Move the hero to the observation tower.
Select the treasure chest as destination with the mouse cursor.
The program displays a default path for the hero.

![](https://i.imgur.com/TvToKng.png)

Click on the treasure chest with the mouse cursor to accept the default path.
At the end of his movement, the hero will not be able to pick up the treasure chest, he will need two more moves.

![](https://i.imgur.com/HmCz6Va.png)

Reload the scenario.
Move the hero to the observation tower.
Move the hero to the sawmill and then to the gazebo.

![](https://i.imgur.com/p7XSNp8.png)

At the end of his movement, the hero will need only one move, instead of two, to be able to pick up the treasure chest.

![](https://i.imgur.com/vKuWslW.png)

This should also be taken into account when the hero can save movement points by passing through interactive locations rather than nearby.
Here again, it is not necessary to change the visual of the hero's path, the movement points are restored in the background and the player does not even notice it.

This enhancement could possibly be optional, but I doubt that even the most masochistic of Heroes 2 purists would not want to benefit from this enhancement. :D

Files: https://drive.google.com/file/d/1pY2cJOATKu67xO_8EXdkr3M-SbHNT0lo/view?usp=sharing

You can read the discussion on this here:
https://heroes2.forumactif.com/t932-enhancement-hero-movement#8443

**Comment by ihhub** [comment]
Hi @LeHerosInconnu , is this issue still reproducible?

**Comment by LeHerosInconnu** [comment]
Hello @ihhub,

> Hi @LeHerosInconnu , is this issue still reproducible?

Yes, and it's worse in the case of the resources, two moves instead of one move to reach the last pile of resources (version 0.9.7 4208).

In fheroes2:

![Hero movement path 001](https://user-images.githubusercontent.com/43583994/135428866-98234a06-2586-4fef-b5a8-debe88d74615.png)

In the case of the path to the Treasure Chest, the default path could go through the active squares of the objects; the interaction window with the active squares would be displayed only if the objects were not already visited (as it is not the initial destination square requested by the player).
In the example, the interaction window of the Sawmill would be displayed to allow the hero to capture it (the green cross on the path should be displayed to indicate the interaction to the player), and the interaction window with the Gazebo would be displayed only if the hero has not visited it yet (if the hero has already visited the Gazebo, the green cross is not displayed and a green arrow is displayed at the location of the Gazebo to indicate to the player that the hero will not stop at the Gazebo).

In fheroes2:

![Hero movement path modifié 002](https://user-images.githubusercontent.com/43583994/135428900-6469aa23-f04f-4c4f-b4a1-c68f8d4f325a.png)

**Comment by oleg-derevenetz** [comment]
Honestly, personally I'm against improvements which are designed to think instead of the player and correct his mistakes retroactively. This game is designed for the correct use of a limited number of heroes and for the construction of logistics chains, and a player who knows how to effectively move heroes gets a well-deserved advantage. IMHO we should not turn this game into a game for monkeys mindlessly pressing a single button - there are already enough of them :)

**Comment by LeHerosInconnu** [comment]
Hello @oleg-derevenetz,

> Honestly, personally I'm against improvements which are designed to think instead of the player and correct his mistakes retroactively. This game is designed for the correct use of a limited number of heroes and for the construction of logistics chains, and a player who knows how to effectively move heroes gets a well-deserved advantage. IMHO we should not turn this game into a game for monkeys mindlessly pressing a single button - there are already enough of them :)

The goal is not to think in the player's place but to avoid him having to do all these micromanagements of moving the heroes.
It is really annoying to have to do this hundreds of times during a game, while validating this way of working the pleasure of the game is increased.
And it would also be good for the AI. :)

I changed the title to include this with two options for my mod:
heroes: hero has no diagonal movement penalty when he performs an action without leaving the square where he is located
heroes: hero's predetermined path passes through the interactive objects to reduce the distance traveled

**Edit.**

Can someone with permission move this thread to "Discussions".
Thanks. :)

**Comment by LeHerosInconnu** [comment]
- "heroes: hero's predetermined path passes through the interactive objects to reduce the distance traveled"
The prediction of the hero's path always goes through the interactive objects of the adventure map to display and follow the shortest possible path.

Here is a simple example of the usefulness of using interactive objects in the hero's path.

In fheroes2:

![Path optimization 001](https://user-images.githubusercontent.com/43583994/200171396-f7ed1691-3cbf-4c6c-bd1f-6e21f9eb34ba.png)

![Path optimization 002](https://user-images.githubusercontent.com/43583994/200171398-2694015b-b007-4fbc-9896-0b1d620a6276.png)

Here is the save file:
[Path to vampires 01.zip](https://github.com/ihhub/fheroes2/files/9945850/Path.to.vampires.01.zip)

**Comment by ihhub** [comment]
Moving this to the Discussion section.
